// eslint-plugin-o1js/build/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findValidContractType = exports.ContractStateTypes = exports.o1jsPrimitiveSizeInfo = exports.geto1jsPrimitive = exports.MAX_CONTRACT_STATES = void 0;
exports.MAX_CONTRACT_STATES = 8;
const o1jsPrimitives = [
    'Field',
    'Bool',
    'UInt32',
    'UInt64',
    'Scalar',
    'PrivateKey',
    'Group',
    'PublicKey',
    'Signature',
];
function geto1jsPrimitive(s) {
    return o1jsPrimitives.find((primitiveType) => s === primitiveType);
}
exports.geto1jsPrimitive = geto1jsPrimitive;
exports.o1jsPrimitiveSizeInfo = {
    Field: { size: 1 },
    Bool: { size: 1 },
    Scalar: { size: 1 },
    UInt32: { size: 1 },
    UInt64: { size: 1 },
    PrivateKey: { size: 1 },
    Group: { size: 2 },
    PublicKey: { size: 2 },
    Signature: { size: 2 },
};
exports.ContractStateTypes = ['prop', 'arrayProp', 'state'];
function findValidContractType(s) {
    return exports.ContractStateTypes.find((type) => s === type);
}
exports.findValidContractType = findValidContractType;
